src/adapter/src/coord/introspection.rs

```diff
@@ -32,6 +32,39 @@ use crate::notice::AdapterNotice;
 use crate::session::Session;
 use crate::{rbac, AdapterError};
 
+fn constant_or_unmaterializable_scalar(scalar: &MirScalarExpr) -> bool {
+    match scalar {
+        mz_expr::MirScalarExpr::Column(_)
+        | mz_expr::MirScalarExpr::Literal(_, _)
+        | mz_expr::MirScalarExpr::CallUnmaterializable(_) => true,
+        mz_expr::MirScalarExpr::CallUnary { .. }
+        | mz_expr::MirScalarExpr::CallBinary { .. }
+        | mz_expr::MirScalarExpr::CallVariadic { .. }
+        | mz_expr::MirScalarExpr::If { .. } => false,
+    }
+}
+
+fn constant_or_unmaterializable(expr: &MirRelationExpr) -> bool {
+    match expr {
+        MirRelationExpr::ArrangeBy { input, keys } => {
+            constant_or_unmaterializable(input)
+                && keys
+                    .iter()
+                    .all(|keys| keys.iter().all(constant_or_unmaterializable_scalar))
+        }
+        MirRelationExpr::Constant { .. } | MirRelationExpr::Get { .. } => true,
+        MirRelationExpr::Map { input, scalars } => {
+            constant_or_unmaterializable(input)
+                && scalars.iter().all(constant_or_unmaterializable_scalar)
+        }
+        MirRelationExpr::Let { value, body, .. } => {
+            constant_or_unmaterializable(value) && constant_or_unmaterializable(body)
+        }
+        MirRelationExpr::Project { input, .. } => constant_or_unmaterializable(input),
+        _ => false,
+    }
+}
```

Comment:
Can you add some comments maybe that just each of these. I don't think I'm familiar enough with `MIR` to know if this is correct.

Reply:
Another option is to pass in the `ResolvedId`s into this and check if the set of IDs is empty after you've filtered out the types and unmaterializable functions.

Reply:
Will add some comments. I can check that the resolved_ids is empty although that should always be true with the current checks. I think I still need the current code to prevent expensive functions

Reply:
I tried that but it doesn't work: unmaterializable functions like now() have an id so resolved_ids isn't empty

Reply:
Can you look up the IDs and filter out the IDs that resolve to an unmaterializable function?

Reply:
added a blurb to the description based on what we slacked earlier. I don't love the solution space (and happy if I'm missing something) but I think this will work.

I believe that even if we do the id work we'll have to walk the expr tree. I cleaned up the code a bit, lemme know what you think